File: code/ch04/4.3.1.SudokuSolver.cpp
class Solution {
private:
    vector<vector<bool>> row_state;
    vector<vector<bool>> column_state;
    vector<vector<bool>> box_state;
    
private:
    void placeNumber(vector<vector<char>>& board, int row, int column, char d) {
        board[row][column] = d;
        row_state[row][d - '1']    = true;
        column_state[column][d - '1'] = true;
        box_state[((row / 3) * 3) + (column / 3)][d - '1'] = true;
    }

    void undoNumberPlace(vector<vector<char>>& board, int row, int column, char d) {
        board[row][column] = '.';
        row_state[row][d - '1'] = false;
        column_state[column][d - '1'] = false;
        box_state[((row / 3) * 3) + (column / 3)][d - '1'] = false;
    }

    void initialState(vector<vector<char>>& board) {
        for (int i = 0; i < 9; i++)
            for (int j = 0; j < 9; j++)
                if (board[i][j] != '.')
                    placeNumber(board, i, j, board[i][j]);
    }
    
    bool recursivePlaceNumber(vector<vector<char>>& board, int row, int column) {
        if (row == 9) return true;
        if (column == 9) return recursivePlaceNumber(board, row + 1, 0);
        if (board[row][column] != '.') return recursivePlaceNumber(board, row, column + 1);
        
        for (int i = '1'; i <= '9'; i++) {
            if (!(row_state[row][i - '1'] 
                  || column_state[column][i - '1'] 
                  || box_state[((row / 3) * 3) + (column / 3)][i - '1'])) {
                placeNumber(board, row, column, i);
                if (recursivePlaceNumber(board, row, column + 1) == true)
                    return true;
                undoNumberPlace(board, row, column, i);
            }
        }
        
        return false;
    }
public:
    void solveSudoku(vector<vector<char>>& board) {
        row_state    = vector<vector<bool>>(9, vector<bool>(9, false));
        column_state = vector<vector<bool>>(9, vector<bool>(9, false));
        box_state    = vector<vector<bool>>(9, vector<bool>(9, false));
        
        initialState(board);
        
        recursivePlaceNumber(board, 0, 0);
    }
};

File: code/ch04/4.3.1.SudokuSolver.py
from typing import List


class Solution:
    row_state = [[False for i in range(10)] for _ in range(9)]
    column_state = [[False for i in range(10)] for _ in range(9)]
    box_state = [[False for i in range(10)] for _ in range(9)]
    board = []

    def solveSudoku(self, board: List[List[str]]) -> None:
        # leetcode 判定时会重复调用函数，因此需要反复初始化状态表
        self.row_state = [[False for i in range(10)] for _ in range(9)]
        self.column_state = [[False for i in range(10)] for _ in range(9)]
        self.box_state = [[False for i in range(10)] for _ in range(9)]
        self.board = board
        for i in range(9):
            for j in range(9):
                if self.board[i][j] != ".":
                    num = int(self.board[i][j])
                    self.row_state[i][num] = True
                    self.column_state[j][num] = True
                    self.box_state[(i // 3) * 3 + j // 3][num] = True

        def recursive_place_number(self, row: int, column: int,) -> bool:
            if column == 9:
                row += 1
                column = 0
                if row == 9:
                    return True

            if self.board[row][column] != ".":
                return recursive_place_number(self, row, column + 1)
            else:
                for i in range(1, 10):
                    if (
                        self.row_state[row][i]
                        or self.column_state[column][i]
                        or self.box_state[(row // 3) * 3 + column // 3][i]
                    ):
                        continue
                    else:
                        self.place_number(row, column, i)
                        if recursive_place_number(self, row, column + 1,):
                            return True
                        self.undo_number_placement(row, column, i)
            return False

        recursive_place_number(self, 0, 0)

    def place_number(self, row: int, column: int, i: int,) -> bool:
        self.row_state[row][i] = True
        self.column_state[column][i] = True
        self.box_state[(row // 3) * 3 + column // 3][i] = True
        self.board[row][column] = str(i)

    def undo_number_placement(self, row: int, column: int, i: int,) -> bool:
        self.row_state[row][i] = False
        self.column_state[column][i] = False
        self.box_state[(row // 3) * 3 + column // 3][i] = False
        self.board[row][column] = "."
